### PR TITLE
Update tools module to support every cell template supported in Cell module

### PR DESCRIPTION
**Pull request by anilbey**
## Updates to `tools` Module

Several methods in the `tools` module were previously missing unit tests and did not reflect the latest updates. This PR addresses these concerns for the following functions:
- `holding_current`
- `detect_spike_step`
- `template_accepts_cvode`
- `calculate_SS_voltage`
- `calculate_input_resistance`
- `detect_hyp_current`
- `detect_spikes`

### Changes Include:
- **Unit Tests:** Added missing tests for better coverage.
- **SONATA Support:** Extended support for modern SONATA cells/circuits.
- **Edge Cases:** Enhanced handling of scenarios that were previously overlooked.
- **Documentation:** Improved docstrings.
- **Type Safety:** Introduced type annotations to functions.

### CODE REVIEW

File: bluecellulab/tools.py [hunk 1]
```diff
@@ -110,51 +113,92 @@ def load_nrnmechanisms(libnrnmech_path: str) -> None:
...
+    reached.
     """
-    cell = bluecellulab.Cell(template_name, morphology_name)
+    cell = bluecellulab.Cell(
+        template_path=template_path,
+        morphology_path=morphology_path,
+        template_format=template_format,
+        emodel_properties=emodel_properties,
+    )
     cell.add_ramp(500, 5000, step_level, step_level)
```
Comment (AurelienJaquier) [inline bluecellulab/tools.py:188]: Do we usually use the ramp protocol for this kind of feature?

**Review by ilkilic — APPROVED** [review]

File: bluecellulab/tools.py (hunk 1 above)
Comment (anilbey) [reply bluecellulab/tools.py:188]: Good point. I had the same question. This is an old code that we did not use for a while.

In this PR I only wanted to add the tests to make sure the function will support the Sonata format and the open-sourced SSCX circuit. I didn't want to touch the protocols.

How about we create an issue from this line and ask Werner why ramp is used here when he's back?

File: bluecellulab/tools.py [hunk 2]
```diff
@@ -213,65 +259,140 @@ def template_accepts_cvode(template_name):
...
+    template_path: str | Path,
+    morphology_path: str | Path,
+    template_format: str,
+    emodel_properties: Optional[EmodelProperties],
+    target_voltage: float,
+    min_current: float,
+    max_current: float,
+) -> float:
     """Search current necessary to bring cell to -85 mV."""
     med_current = min_current + abs(min_current - max_current) / 2
```
[the remark this answers is not in the document]
Comment (anilbey) [reply bluecellulab/tools.py:272]: I don't know, someone else wrote this :))

File: bluecellulab/tools.py (hunk 1 above)
Comment (AurelienJaquier) [reply bluecellulab/tools.py:188]: alright, sounds good

File: bluecellulab/tools.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (AurelienJaquier) [reply bluecellulab/tools.py:272]: :))

**Review by AurelienJaquier — APPROVED** [review]
looks great!

File: bluecellulab/tools.py (hunk 1 above)
Comment (anilbey) [reply bluecellulab/tools.py:188]: Here it is: https://github.com/BlueBrain/BlueCelluLab/issues/77

Comment (AurelienJaquier) [reply bluecellulab/tools.py:188]: thanks!
